File: python/mirage/vfs/ssh/ssh.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.accessor.ssh import SSHAccessor
from mirage.commands.builtin.ssh import COMMANDS as SSH_COMMANDS
from mirage.commands.builtin.ssh.io import IO
from mirage.core.ssh.config import SSHConfig
from mirage.core.ssh.watch import build_delta_hook
from mirage.ops.ssh import OPS as SSH_OPS
from mirage.types import VFSName
from mirage.vfs.bound import BoundVFS
from mirage.vfs.ssh.prompt import PROMPT
from mirage.watch.base import DeltaHook


class SSHVFS(BoundVFS):

    accessor: SSHAccessor
    name: str = VFSName.SSH
    caches_reads: bool = True
    # SFTP stat/readdir report the remote inode's exact byte size for
    # every file; reads are the same raw bytes.
    SIZES_ALWAYS_KNOWN: bool = True
    # A remote filesystem: short-lived index, long enough to spare a
    # re-walk inside one command pipeline. Mirrors the TypeScript VFS.
    index_ttl: float = 60
    PROMPT: str = PROMPT

    def __init__(self, config: SSHConfig) -> None:
        super().__init__(io=IO)
        self.config = config
        self.accessor = SSHAccessor(self.config)
        for fn in SSH_COMMANDS:
            self.register(fn)
        for ro in SSH_OPS:
            self.register_op(ro)

    def delta_hook(self) -> DeltaHook:
        return build_delta_hook(self.accessor)

    def get_state(self) -> dict[str, Any]:
        return self.config_state(self.config)

File: typescript/packages/node/src/vfs/ssh/ssh.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { BoundVFS } from '@struktoai/mirage-core/vfs/bound'
import { SSH_IO } from '../../commands/builtin/ssh/io.ts'

import type { RegisteredCommand } from '@struktoai/mirage-core/commands/config'
import type { RegisteredOp } from '@struktoai/mirage-core/ops/registry'

import type { FindOptions, VFS } from '@struktoai/mirage-core/vfs/base'
import type { PathSpec } from '@struktoai/mirage-core/types'
import { VFSName } from '@struktoai/mirage-core/types'

import { SSHAccessor } from '../../accessor/ssh.ts'
import { SSH_COMMANDS } from '../../commands/builtin/ssh/index.ts'

import { SSH_OPS } from '../../ops/ssh/index.ts'
import { type SSHConfig, type SSHConfigRedacted, redactSshConfig } from './config.ts'
import { SSH_PROMPT } from './prompt.ts'
import { type DeltaHook } from '@struktoai/mirage-core/watch/index'
import { buildDeltaHook } from '../../core/ssh/watch.ts'

export interface SSHVFSState {
  type: string
  config: SSHConfigRedacted
}

export class SSHVFS extends BoundVFS<SSHAccessor> implements VFS {
  declare writeFile: (p: PathSpec, data: Uint8Array) => Promise<void>

  declare appendFile: (p: PathSpec, data: Uint8Array) => Promise<void>

  declare exists: (p: PathSpec) => Promise<boolean>

  declare mkdir: (p: PathSpec, options?: { recursive?: boolean }) => Promise<void>

  declare rmdir: (p: PathSpec) => Promise<void>

  declare unlink: (p: PathSpec) => Promise<void>

  declare rename: (src: PathSpec, dst: PathSpec) => Promise<void>

  declare truncate: (p: PathSpec, length: number) => Promise<void>

  declare copy: (src: PathSpec, dst: PathSpec) => Promise<void>

  declare rmR: (p: PathSpec) => Promise<void>

  declare du: (p: PathSpec) => Promise<number>

  declare find: (p: PathSpec, options?: FindOptions) => Promise<string[]>

  readonly kind = VFSName.SSH
  readonly cachesReads: boolean = true
  // SFTP stat/readdir report the remote inode's exact byte size for every
  // file; reads are the same raw bytes.
  readonly sizesAlwaysKnown: boolean = true
  override readonly indexTtl: number = 60
  readonly prompt = SSH_PROMPT
  readonly config: SSHConfig
  readonly accessor: SSHAccessor

  constructor(config: SSHConfig) {
    super(SSH_IO)
    this.config = config
    this.accessor = new SSHAccessor(config)
  }

  override async close(): Promise<void> {
    await this.accessor.close()
    await super.close()
  }

  ops(): readonly RegisteredOp[] {
    return SSH_OPS
  }

  commands(): readonly RegisteredCommand[] {
    return SSH_COMMANDS
  }

  deltaHook(): DeltaHook {
    return buildDeltaHook(this.accessor)
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  override async getState(): Promise<SSHVFSState> {
    return {
      type: this.kind,
      config: redactSshConfig(this.config),
    }
  }
}
